internal/dag/seq.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package dag
+
+import (
+	"iter"
+	"slices"
+)
+
+type VertexSeq[T Vertex] iter.Seq[T]
+
+func (seq VertexSeq[T]) Collect() []T {
+	return slices.Collect(iter.Seq[T](seq))
+}
+
+// Vertices returns an iterator over all the vertices in the graph.
+func (g *Graph) VerticesSeq() VertexSeq[Vertex] {
+	return func(yield func(v Vertex) bool) {
+		for _, v := range g.vertices {
+			v, ok := v.(Vertex)
+			if !ok {
+				continue
+			}
+			if !yield(v) {
+				return
+			}
+		}
+	}
+}
+
+// SelectSeq filters a sequence to include only elements that can be type-asserted to type U.
+// It returns a new sequence containing only the matching elements.
+// The yield function can return false to stop iteration early.
+func SelectSeq[T Vertex, U Vertex](seq VertexSeq[T], filter func(U)) VertexSeq[U] {
```

Comment:
I think we don't need the filter function at all here? It's not being called, and you can specify the type using the `[...]` syntax: `SelectSeq[MockVertex2, MockVertex](...)`.